wave_lang/kernel/wave/codegen/read_write.py

```diff
@@ -747,179 +738,260 @@ def handle_tensor_load_to_lds(emitter: WaveEmitter, node: fx.Node):
-    dim_stride_1_trunc = lshift(dim_stride_1_lower, 16)
-    pack = arith_d.ori(dim_stride_0_trunc, dim_stride_1_trunc)
-    d1 = vector_d.insert(pack, d1, static_position=[6], dynamic_position=[])
+        local_bounds = [
+            bounds[s] - global_tile_index_current[s].start for s in symbolic_shape
+        ]
+        local_bounds = [gen_sympy_index(subs, b) for b in local_bounds]
 
-    # shift dim stride 1 to get upper 32 bit and pack to i32
-    dim_stride_1_sh = rshift(dim_stride_1, 16)
-    pack = arith_d.trunci(i32, dim_stride_1_sh)
-    d1 = vector_d.insert(pack, d1, static_position=[7], dynamic_position=[])
+        strides = strides_from_symbolic_shape(
+            IndexingContext.current(), symbolic_shape, allow_mixed_shapes=True
+        )
+        # Descriptor assumes rightmost stride 1 and expect last stride as full data size
+        strides = [strides[0] * symbolic_shape[0]] + strides[:-1]
+        strides = [gen_sympy_index(subs, s) for s in strides]
+
+        distributed_shape_vals = [
+            gen_sympy_index(subs, distributed_shape[s]) for s in symbolic_shape
+        ]
+
+        d0 = vector_d.broadcast(vec_type_4, c0)
+        d1 = vector_d.broadcast(vec_type_8, c0)
+        d2 = vector_d.broadcast(vec_type_4, c0)
+        d3 = vector_d.broadcast(vec_type_4, c0)
+
+        # descriptor properties
+        mode = 2  # vimage
+        valid = 1
+        dim_stride_1 = arith_d.index_cast(i48, strides[0])
+        dim_stride_0 = arith_d.index_cast(i48, strides[1])
+        tile_size_1 = arith_d.index_cast(i32, distributed_shape_vals[0])
+        tile_size_0 = arith_d.index_cast(i32, distributed_shape_vals[1])
+        dim_size_1 = arith_d.index_cast(i32, local_bounds[0])
+        dim_size_0 = arith_d.index_cast(i32, local_bounds[1])
+
+        # 0: 1 byte; 1: 2 byte; 2: 4 byte; 3: 8 byte
+        descriptor_type = lambda x: int(math.log2(x.bitwidth() >> 3))
+        data_size = cast_py_value(emitter, descriptor_type(element_type), i32).ir_value
+
+        global_mem = cast_py_value(emitter, src)
+        shared_mem = cast_py_value(emitter, dst)
+
+        global_value = global_mem.ir_value
+        shared_value = shared_mem.ir_value
+
+        bytewidth = element_type.bitwidth() // 8
+        element_byte_index = arith_d.constant(IndexType.get(), bytewidth)
+
+        # calculcate global address
+        # 0. breakdown index sequence to WG & TH offsets : ele
+        # 1. uniform per wave access : ele
+        # 2. linearize global memory buffer
+        # 3. offset = X + Y * tensor dim 0 stride : ele
+        # 4. offset_byte = offset * element byte : byte
+        # 5. get global memory pointer
+        # 6. move global memory pointer by offset_byte to get global address of a tile : byte
+        index, _, _ = _build_start_indices(emitter, global_tile_index_current)
+
+        wave_index_x = assume_index_subgroup_uniform(index[1], i32)  # k
+        wave_index_y = assume_index_subgroup_uniform(index[0], i32)  # m
```

Comment:
slightly unrelated but would be nice to have this as a util function that actually checks for subgroup uniform-ness within index and use https://mlir.llvm.org/docs/Dialects/ROCDLDialect/#rocdlreadfirstlane-rocdlreadfirstlaneop instead of the llvm_d. Actually would be really nice to have one clean-up PR to convert existing old llvm dialect intrinsic calls as their new rocdl form :)